router/usecase/candidate_routes.go

```diff
@@ -141,6 +143,136 @@ func GetCandidateRoutes(pools []sqsdomain.PoolI, tokenIn sdk.Coin, tokenOutDenom
 	return validateAndFilterRoutes(routes, tokenIn.Denom, logger)
 }
 
+// GetCandidateRoutesNew new algorithm for demo purposes.
+// Note: implementation is for demo purposes and is to be further optimized.
+// TODO: spec, unit tests via https://linear.app/osmosis/issue/DATA-250/[candidaterouteopt]-reimplement-and-test-getcandidateroute-algorithm
+func GetCandidateRoutesNew(poolsByDenom map[string][]sqsdomain.PoolI, tokenIn sdk.Coin, tokenOutDenom string, maxRoutes, maxPoolsPerRoute int, logger log.Logger) (sqsdomain.CandidateRoutes, error) {
```

Comment:
Note to reviewer: this is mostly a copy of the old ` GetCandidateRoutes` implementation above to put things into perspective.

See that the first parameter differs.

Now, instead of reiterating all pools, we only need to iterate over pools that contain the current token in.

Reply:
🆗